instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/EVRouteRefreshTest.kt

```diff
@@ -0,0 +1,688 @@
+
+        assertEquals(
+            listOf(29, 13),
+            requestedRoutes[0].getSocAnnotationsFromLeg(0)!!.firstLastAnd()
+        )
+        assertEquals(
+            listOf(43, 10),
+            requestedRoutes[0].getSocAnnotationsFromLeg(1)!!.firstLastAnd()
+        )
+        assertEquals(
+            listOf(null, 8097, null),
+            requestedRoutes[0].directionsResponse.waypoints()?.extractChargeAtArrival()
+        )
+
+        assertEquals(
+            listOf(28, 12),
+            updatedRoutes[0].getSocAnnotationsFromLeg(0)!!.firstLastAnd()
+        )
+        assertEquals(
+            listOf(42, 10),
+            updatedRoutes[0].getSocAnnotationsFromLeg(1)!!.firstLastAnd()
+        )
+        assertEquals(
+            listOf(null, 7286, null),
+            updatedRoutes[0].directionsResponse.waypoints()?.extractChargeAtArrival()
+        )
+    }
+
+    @Test
+    fun ev_route_refresh_updates_ev_annotations_and_waypoints_for_second_leg() = sdkTest {
+        val routeGeometryIndex = 1050
+        val legGeometryIndex = 300
+        addRefreshRequestHandler(
+            R.raw.ev_route_refresh_response_for_second_leg,
+            acceptedGeometryIndex = routeGeometryIndex
+        )
+        val requestedRoutes = requestRoutes(twoCoordinates, electric = true)
+        mapboxNavigation.setEVDataUpdater(evDataUpdater)
+        val evData = mapOf(
+            KEY_ENERGY_CONSUMPTION_CURVE to "0,300;20,160;80,140;120,180",
+            KEY_EV_INITIAL_CHARGE to "17000",
+            KEY_EV_PRECONDITIONING_TIME to "10",
+            KEY_AUXILIARY_CONSUMPTION to "300"
+        )
+        evDataUpdater.updateData(evData)
+        mapboxNavigation.setNavigationRoutes(requestedRoutes, initialLegIndex = 1)
+        // corresponds to currentRouteGeometryIndex = 1050
+        stayOnPosition(48.435946, 10.86999)
+        mapboxNavigation.startTripSession()
+        mapboxNavigation.routeProgressUpdates().filter { progress ->
+            progress.currentRouteGeometryIndex == routeGeometryIndex
+        }.first()
+
+        val updatedRoutes = waitUntilRefresh().navigationRoutes
+
+        assertEquals(
+            listOf(29, 13),
+            requestedRoutes[0].getSocAnnotationsFromLeg(0)!!.firstLastAnd()
+        )
+        assertEquals(
+            listOf(43, 38, 10),
+            requestedRoutes[0].getSocAnnotationsFromLeg(1)!!.firstLastAnd(legGeometryIndex)
+        )
+        assertEquals(
+            listOf(null, 8097, null),
+            requestedRoutes[0].directionsResponse.waypoints()?.extractChargeAtArrival()
+        )
+
+        assertEquals(
+            listOf(29, 13),
+            updatedRoutes[0].getSocAnnotationsFromLeg(0)!!.firstLastAnd()
+        )
+        assertEquals(
+            listOf(43, 28, 1),
+            updatedRoutes[0].getSocAnnotationsFromLeg(1)!!.firstLastAnd(legGeometryIndex)
+        )
+        assertEquals(
+            listOf(null, 8097, null),
```

Comment:
This check fails for now because I've used the real Directions API response to form our test response but it returned less waypoints than I expected (and than as described in their docs). I've asked Mandeep a corresponding question.

Reply:
The fixed a bug, I've updated the mocked response appropriately. Should be working now.